### PR TITLE
[Shared dev] [#162254] Open "Report an issue" in modal

### PR DESCRIPTION
**Pull request by LeticiaErrandonea**
# Release Notes
* Show "Report an issue" action in modal

# Screenshot
image

# Accessibility
- [ ] Did you scan for accessibility issues?
- [ ] Did you check our accessibility goal checklist?
- [ ] Did you add accessibility [specs](https://github.com/dequelabs/axe-core-gems/blob/develop/packages/axe-core-rspec/README.md)?

### CODE REVIEW

File: app/assets/stylesheets/app/order_management.scss [hunk 1]
```diff
@@ -28,6 +28,10 @@
   top: 6px;
 }
 
+#js--reportAnIssueModal {
```
Comment (LeticiaErrandonea) [inline app/assets/stylesheets/app/order_management.scss:31]: More styles are needed as modal still looks off. Also, I think it's missing a Cancel button

image

File: app/controllers/instrument_issues_controller.rb [hunk 2]
```diff
@@ -10,13 +10,24 @@ class InstrumentIssuesController < ApplicationController
   before_action :init_instrument_issue, only: %i[new create]
 
   def new
+    redirect_to_order_id = params[:redirect_to_order_id]
```
Comment (giladshanan) [inline app/controllers/instrument_issues_controller.rb:13]: try using the `modal?` pattern, at least for consistency

File: app/presenters/reservation_user_action_presenter.rb [hunk 3]
```diff
@@ -131,7 +133,12 @@ def move_link
   end
 
   def report_an_issue_link
-    link_to(I18n.t("views.instrument_issues.new.title"), new_facility_order_order_detail_issue_path(facility, order, order_detail))
+    if @redirect_to_order_id.present?
```
Comment (giladshanan) [inline app/presenters/reservation_user_action_presenter.rb:136]: Consider wrapping this in a method for documentation and clarity like: `modal_display?`

File: app/controllers/instrument_issues_controller.rb [hunk 4]
```diff
@@ -8,15 +8,27 @@ class InstrumentIssuesController < ApplicationController
...
+    end
   end
 
   def create
     @instrument_issue.assign_attributes(create_params)
+    redirect_to_order_id = params[:redirect_to_order_id]
 
     if @instrument_issue.send_notification
-      redirect_to reservations_path, notice: text("create.success")
+      if redirect_to_order_id.present?
```
Comment (giladshanan) [inline app/controllers/instrument_issues_controller.rb:26]: let's move this into a method that returns the redirect path
in that method we can check modal? and ensure we have an order to redirect to 
we can also add a code comment there if needed to clarify the expected code paths

File: app/presenters/reservation_user_action_presenter.rb [hunk 5]
```diff
@@ -142,4 +149,8 @@ def can_fix_problem?
     OrderDetails::ProblemResolutionPolicy.new(order_detail).user_can_resolve?
   end
 
+  def modal_display?
```
Comment (giladshanan) [inline app/presenters/reservation_user_action_presenter.rb:152]: Add a comment here to describe the use case/code path
when/why/which feature?

File: app/assets/stylesheets/app/order_management.scss (hunk 1 above)
Comment (giladshanan) [reply app/assets/stylesheets/app/order_management.scss:31]: we have existing examples on the code, or you can look at bootstrap 2 docs

**Review by giladshanan — COMMENTED** [review]
left some comments

File: app/controllers/instrument_issues_controller.rb [hunk 6]
```diff
@@ -8,15 +8,21 @@ class InstrumentIssuesController < ApplicationController
...
   before_action :init_product
   before_action :init_instrument_issue, only: %i[new create]
+  layout -> { modal? ? false : "application" }
 
   def new
+    @redirect_to_order_id = params[:redirect_to_order_id]
+
+    if modal?
+      render layout: false
+    end
```
Comment (giladshanan) [inline app/controllers/instrument_issues_controller.rb:18]: I think line 11 should handle this

File: app/controllers/instrument_issues_controller.rb [hunk 7]
```diff
@@ -37,9 +43,29 @@ def init_product
...
-    @instrument_issue = InstrumentIssue.new(product: @product, 
+    @instrument_issue = InstrumentIssue.new(product: @product,
                                             user: current_user,
                                             order_detail: @order_detail)
   end
 
+  def redirect_to_path
+    redirect_to_order_id = params[:redirect_to_order_id]
+
+    if redirect_to_order_id.present?
```
Comment (giladshanan) [inline app/controllers/instrument_issues_controller.rb:54]: should we be checking `modal?` here?  
can you add a comment: `# this means the request came from a modal on the order show page`

File: app/controllers/instrument_issues_controller.rb [hunk 8]
```diff
@@ -37,9 +43,29 @@ def init_product
...
+      order = Order.find(redirect_to_order_id)
+
+      if order.present?
+        facility_order_path(order.facility, order)
+      else
+        reservations_path
```
Comment (giladshanan) [inline app/controllers/instrument_issues_controller.rb:60]: do we expect to be hitting this?  maybe add a comment to say when we would hit it, or that we're just putting it in for safety

File: app/views/instrument_issues/new.html.haml [hunk 9]
```diff
@@ -1,3 +1,5 @@
+- modal_view = @redirect_to_order_id.present?
```
Comment (giladshanan) [inline app/views/instrument_issues/new.html.haml:1]: can we set this in the controller?

File: app/views/instrument_issues/new.html.haml [hunk 10]
```diff
@@ -9,15 +11,28 @@
 = content_for :h1 do
   = @product.name
 
-.row
-  .span12
-    %h2= text("title")
+- url = facility_order_order_detail_issues_path(current_facility, @order_detail.order, @order_detail, redirect_to_order_id: @redirect_to_order_id)
```
Comment (giladshanan) [inline app/views/instrument_issues/new.html.haml:14]: can we set this in the controller?

File: app/views/instrument_issues/new.html.haml [hunk 11]
```diff
@@ -9,15 +11,28 @@
...
+    .modal-header
+      = modal_close_button
+      %h2= text("title")
+      %p= text("instructions")
+    .modal-body
+      = simple_form_for @instrument_issue, url: url, method: :post do |f|
+        = f.input :message, as: :text, input_html: { class: "input-xxlarge" }
+        = f.submit text("submit"), class: "btn btn-primary"
+    .modal-footer
+      = modal_cancel_button text: "Cancel"
```
Comment (giladshanan) [inline app/views/instrument_issues/new.html.haml:27]: i18n please

File: spec/support/contexts/cross_core_context.rb [hunk 12]
```diff
@@ -36,8 +36,9 @@
       create(:purchased_order, cross_core_project: cross_core_project2, product: item, account: accounts.last),
       create(:purchased_order, cross_core_project: cross_core_project2, product: facility3_item, account: accounts.last),
       # cross_core_project3 has no order details from facility 1
-      create(:purchased_order, cross_core_project: cross_core_project3, product: facility2_item2, account: accounts.last),
+      create(:purchased_order, cross_core_project: cross_core_project3, product: facility2_instrument, account: accounts.last),
     ]
   end
 
+  let(:cross_core_reservation_order) { create(:setup_order, cross_core_project:, product: facility2_instrument, account: accounts.last) }
```
Comment (giladshanan) [inline spec/support/contexts/cross_core_context.rb:43]: why do we need this one?  can you add a comment to explain?

File: spec/system/admin/reservation_actions_spec.rb [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    allow_any_instance_of(InstrumentIssue).to receive(:send_notification).and_return(true)
+
+    login_as facility_administrator
+    visit facility_order_path(facility, originating_order_facility1)
+  end
+
+  describe "Report an Issue" do
+    it "redirects to original order show" do
+      find("h3", text: cross_core_reservation_order.facility.to_s, match: :first).click
+      find("a", text: "Report an Issue").click
```
Comment (giladshanan) [inline spec/system/admin/reservation_actions_spec.rb:20]: maybe put this into a `within` block to make it explicit which order detail we're reporting an issue about

**Review by giladshanan — COMMENTED** [review]
Nice work on the refactors!

File: app/controllers/instrument_issues_controller.rb (hunk 7 above)
Comment (LeticiaErrandonea) [reply app/controllers/instrument_issues_controller.rb:54]: The request for `create` returns `false` for `request.xhr?`, so I had to remove the `modal?` condition from here.

**Review by giladshanan — APPROVED** [review]

**Pull request merged**